File: leetcode/dp/279. Perfect Squares/py/queue.py
from collections import deque

class Solution:
    def numSquares(self, n):
      squareNums = [i * i for i in range(1, int(n**0.5)+1)]
      
      cnt = 0
      queue = deque()
      seen = set()

      queue.append(n)

      while len(queue):
        s = len(queue)

        for _ in range(s):
          num = queue.popleft()

          if num == 0:
            return cnt

          for square in squareNums:
            if square > num:
              break

            remainder = num - square
            if remainder >= 0 and remainder not in seen:
              seen.add(remainder)
              queue.append(remainder)
        cnt += 1
      return -1

File: leetcode/dp/279. Perfect Squares/js/queue.js
const numSquares = target => {
  let squareNums = new Set();

  for (let i = 1; i * i <= target; i++) {
    squareNums.add(i * i);
  }

  let cnt = 0;
  let queue = [];
  let seen = new Set();

  queue.push(target);

  while (queue.length) {
    let size = queue.length;

    for (let i = 0; i < size; i++) {
      let num = queue.shift();

      if (num == 0) return cnt;

      for (let square of squareNums.values()) {
        if (square > num) break;

        if (num - square >= 0 && !seen.has(num - square)) {
          queue.push(num - square);
          seen.add(num - square);
        }
      }
    }
    cnt++;
  }

  return -1;
};
